maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt

```diff
@@ -0,0 +1,72 @@
+package org.odk.collect.maps.layers
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import org.odk.collect.androidshared.livedata.MutableNonNullLiveData
+import org.odk.collect.androidshared.livedata.NonNullLiveData
+import org.odk.collect.async.Scheduler
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.settings.keys.ProjectKeys
+
+class OfflineMapLayersPickerViewModel(
```

Comment:
I need this view model scoped to its fragment not activity so that the data is cleared when the bottom sheet is closed.
The data I want to clear is:
`val existingLayers: LiveData<List<CheckableReferenceLayer>> = _existingLayers`
in `CheckableReferenceLayer` I keep booleans to know if a layer is expanded/collapsed or temporarily selected (without confirming by clicking the save button). If it wasn't cleared after closing and opening the bottom sheet again the state would be preserved which should not happen.
It also make more sense to have this separate view model for data and functions that are not shared,

Reply:
I'd argue that the data is shared here: one view (the picker) displays the layers, while the other (the importer) can modify them forcing the first one to update. This means we still need to have some trigger between the two. 

The current implementation uses a listener on the Activity level view model that points at the Fragment level view model. This will unfortunately create a memory leak  due to the different lifecycles going on. You can trigger this by rotating the device with the picker up a couple of times:

<img src="https://github.com/getodk/collect/assets/556280/48f5a6bd-e2b4-4961-b015-b05516fa8930" width="320px"/>

This is because the `OfflineMapLayersSharedViewModel` instance owned by the Activity will end up with an implicit reference to each `OfflineMapLayersPicker` instance through the block referencing the Fragment's instance of `OfflineMapLayersPickerViewModel` (through the `viewModel` field).

There are a few solutions I can think of here:

1. Move the shared data to a shared place like a new layers "data service". This would allow us to share a reactive data value (like `Flow`/`LiveData`) between view models and encapsulate data modifications like we do for forms/instances already.
2. Keep the old `OfflineMapLayersViewModel` as it was, and then wrap its data (`#existingLayers`) in another `ViewModel` that is responsible for selection state that can live at the Fragment level. We already have a framework for doing this with multi selects (`MultiSelectViewModel`) and I reckon the single select version would be pretty similar.
3. Just hold the selection state in the Fragment itself and either don't worry about the state being cleared on rotation, or just use saved UI state.

I'd personally opt for option 2. 3 feels like something we might need to clean up later and 1 is probably not something we need yet (as we're not sharing the layers data across many places).

And just a quick note on the general architecture here: it's easy to get into debates around the aesthetics of different structures, but one big advantage of trying to adopt a more ["unidirectional data flow" approach](https://developer.android.com/topic/architecture/ui-layer#udf) (which is the closest generalized description of what I'm going for here in my opinion) is that it's really helpful for avoiding memory leaks and the even worse "dead object" based problems (like an AsyncTask trying to call an Activity that doesn't exist) as it keeps knowledege of the view away from as much as possible.

Reply:
Option 2 is basically what I have now, isn't it? Ok, the listener approach was a bad solution but I have moved `#existingLayers` to a separate ViewModel scoped to its fragment.
The only difference is that I moved there the method responsible for loading existing layers (and saving the selected layer) as according to what I said above it's not used by the import so it doesn't make sense to share it. Don't you like that I moved that non-shared code away from the shared view model? Would you prefer to still keep everything in one ViewModel and only have a separate one for the data?

Reply:
> Option 2 is basically what I have now, isn't it? Ok, the listener approach was a bad solution but I have moved #existingLayers to a separate ViewModel scoped to its fragment.

Yeah, I think option 2 is definitely the closest to the last and this PR. The main difference in what I was proposing was that all the "layers" data lives in one shared (Activity level) ViewModel (like we had before), and the selection/expanded state lives in a generic Fragment level ViewModel that works like `MultiSelectViewModel` rather than having a Fragment level ViewModel that deals with both layers and selection/toggle state. 

Although more different structurally, it might actually be that option 1 (introducing an `AppState` backed data service for existing layers) is more ideal here. That'd make the whole thing simpler as it'd support having two ViewModels (one focused on existing layers and one focused on importing like we have here) without needing any direct interaction between them. I'd still prefer to have a generic way to do the single select/expandable list, but that's something we could always extract later.

Reply:
I would like to avoid the first option because:
1. It seems to be too complex for what we have here
2. I would need to clear the state either way on dismissing the bottom sheet.

Is this https://github.com/getodk/collect/pull/6175/commits/3b1e94ced72f264cd5ce0265b199152b94a3aedf what you were thinking about when it comes to option 2?

Reply:
I'll wait to see the new stuff you were mentioning in our Slack call to comment! The main thing I was going to say was that it would feel nicer to me to have `OfflineMapLayersStateViewModel` focus on the checked/expanded state (instead of also need to hold the whole list of layers) and let `OfflineMapLayersViewModel` focus on the layers data. Just to get it out of my head I was imagining a generic `SelectExpandViewModel` (terrible name) with an interface like:

```kotlin
class SelectExpandViewModel : ViewModel() {

    fun getSelected(): LiveData<String>
    fun getExpanded(): LiveData<String?>

    fun select(id: String)
    fun toggleExpanded(id: String)
}
```

The Fragment could then just observe `getSelected`/`getExpanded` and set them on the list adapter when they change and the list items click listeners could call `select`/`toggleExpanded`. Again, this is a similar approach to `MultiSelectViewModel`.

Reply:
>  was imagining a generic SelectExpandViewModel (terrible name) 

I was also having a hard time trying to come up with a good name so for now I kept the old one.